lib/Adapter/Tolerant/TolerantIndexBuilder.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+namespace Phpactor\Indexer\Adapter\Tolerant;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\QualifiedName;
+use Microsoft\PhpParser\Node\Statement\ClassDeclaration;
+use Microsoft\PhpParser\Node\Statement\FunctionDeclaration;
+use Microsoft\PhpParser\Parser;
+use Phpactor\Indexer\Model\Index;
+use Phpactor\Indexer\Model\IndexBuilder;
+use Phpactor\Indexer\Model\Record\ClassRecord;
+use Phpactor\Indexer\Model\Record\FunctionRecord;
+use Phpactor\Name\FullyQualifiedName;
+use Phpactor\TextDocument\ByteOffset;
+use SplFileInfo;
+
+class TolerantIndexBuilder implements IndexBuilder
+{
+    /**
+     * @var Index
+     */
+    private $index;
+
+    /**
+     * @var Parser
+     */
+    private $parser;
+
+    public function __construct(
+        Index $index,
+        ?Parser $parser = null
+    ) {
+        $this->index = $index;
+        $this->parser = $parser ?: new Parser();
+    }
+
+    public function index(SplFileInfo $info): void
+    {
+        $contents = @file_get_contents($info->getPathname());
+
+        if (false === $contents) {
+            return;
+        }
+
+        $node = $this->parser->parseSourceFile($contents, $info->getPathname());
+        $this->indexNode($info, $node);
+    }
+
+    public function done(): void
+    {
+        $this->index->updateTimestamp();
+    }
+
+    private function indexNode(SplFileInfo $info, Node $node): void
+    {
+        if ($node instanceof ClassDeclaration) {
+            $this->indexClassDeclaration($info, $node);
+            return;
+        }
+
+        if ($node instanceof FunctionDeclaration) {
+            $this->indexFunction($info, $node);
+            return;
+        }
+
+        foreach ($node->getChildNodes() as $childNode) {
+            $this->indexNode($info, $childNode);
+        }
+    }
+
+    private function indexClassDeclaration(SplFileInfo $info, ClassDeclaration $node): void
+    {
+        $record = $this->index->get(ClassRecord::fromName($node->getNamespacedName()->getFullyQualifiedNameText()));
+        assert($record instanceof ClassRecord);
+        $record->withLastModified($info->getCTime());
+        $record->withStart(ByteOffset::fromInt($node->getStart()));
+        $record->withType('class');
+        $record->withFilePath($info->getPathname());
```

Comment:
with => set (`with` kindof hints at immutability)

Reply:
Absolutely!
Maybe also use a fluent style to improve readability.

Reply:
I just saw the `Record` class:
Given all information are accepted in the constructor you could use it instead.
Or even make `Record` an immutable value object, then you ca decide if you keep the method (if used anywhere else) or remove them alltogether.

Reply:
Having an immutability on `Record` would be nice IMO

Reply:
Not sure it makes sense - it _is_ mutable and is _expected_ be an invalid state (for optimization) - hopefully it's eventaully valid, but it's not guranteed.

Currently parsing a class which references another class will create a record for the reference (in order to add a reference to it) but that reference may never be indexed itself.